utils/kubernetes/service.go

```diff
@@ -2,109 +2,74 @@ package kubernetes
+	obj, err := client.CoreV1().Services(opts.Namespace).Get(ctx, opts.Name, metav1.GetOptions{})
 	if err != nil {
 		return nil, ErrServiceDiscovery(err)
 	}
-	if endpoint := extractNodePortEndpoint(svc, nodes); endpoint != nil {
-		return endpoint, nil
-	}
 
-	// Try clusterip endpoint
-	if endpoint := extractClusterIPEndpoint(svc); endpoint != nil {
-		return endpoint, nil
-	}
-
-	return nil, err
+	return GetEndpoint(ctx, opts, obj)
 }
 
-// extractLoadBalancerEndpoint extracts loadbalancer based endpoint, if any.
-// It returns the nil if no valid endpoint is extracted
-func extractLoadBalancerEndpoint(svc *coreV1.Service) *utils.Endpoint {
-	ports := svc.Spec.Ports
-	ingresses := svc.Status.LoadBalancer.Ingress
+func GetEndpoint(ctx context.Context, opts *ServiceOptions, obj *corev1.Service) (*utils.Endpoint, error) {
+	var nodePort, clusterPort int32
+	endpoint := utils.Endpoint{}
 
-	for _, ingress := range ingresses {
-		var address string
-		if ingress.Hostname != "" && ingress.Hostname != "None" {
-			address = ingress.Hostname
-		} else if ingress.IP != "" {
-			address = ingress.IP
-		} else {
-			// If no valid ip address and hostname is found
-			// then move on to the next ingress
-			continue
+	if opts.PortNumber == nil {
+		for _, port := range obj.Spec.Ports {
+			if port.Name == opts.PortSelector {
+				nodePort = port.NodePort
+				clusterPort = port.Port
+			}
 		}
-
-		for _, port := range ports {
-			return &utils.Endpoint{
-				Name:    svc.GetName(),
-				Address: address,
-				Port:    port.Port,
+	} else {
+		for _, port := range obj.Spec.Ports {
+			if &port.Port == opts.PortNumber {
+				nodePort = port.NodePort
+				clusterPort = port.Port
 			}
 		}
 	}
 
-	return nil
-}
-
-// extractNodePortEndpoint extracts nodeport based endpoint, if any.
-// It returns the nil if no valid endpoint is extracted
-func extractNodePortEndpoint(svc *coreV1.Service, nl *coreV1.NodeList) *utils.Endpoint {
-	ports := svc.Spec.Ports
-
-	for _, node := range nl.Items {
-		for _, addressData := range node.Status.Addresses {
-			if addressData.Type == "InternalIP" {
-				address := addressData.Address
-
-				for _, port := range ports {
-					// nodeport 0 is an invalid nodeport
-					if port.NodePort != 0 {
-						return &utils.Endpoint{
-							Name:    svc.GetName(),
-							Address: address,
-							Port:    port.NodePort,
-						}
-					}
-				}
-			}
-		}
+	// get clusterip endpoint
+	endpoint.Internal = &utils.HostPort{
+		Address: obj.Spec.ClusterIP,
+		Port:    clusterPort,
 	}
-	return nil
-}
 
-// extractClusterIPEndpoint extracts clusterIP based endpoint, if any.
-// It returns the nil if no valid endpoint is extracted
-func extractClusterIPEndpoint(svc *coreV1.Service) *utils.Endpoint {
-	ports := svc.Spec.Ports
-	clusterIP := svc.Spec.ClusterIP
+	endpoint.External = &utils.HostPort{
+		Address: "localhost",
```

Comment:
Will "localhost" always be the correct reference?

Reply:
its the default value.